panic: runtime error: index out of range

Condition: 
go version go1.11.6 linux/arm
cat top10milliondomains.txt | ~/go/bin/diosts -t 100 2>diosts.log >securitytxt.json
Source: https://www.domcop.com/files/top/top10milliondomains.csv.zip

Expected result: 
Completion of task

Result:
```
goroutine 116 [running]:
github.com/disclose/diosts/pkg/securitytxt.baseDomain(0x2706707, 0x12, 0x12, 0x7)
	/home/pi/go/src/github.com/disclose/diosts/pkg/securitytxt/domain.go:209 +0x10c
github.com/disclose/diosts/pkg/securitytxt.checkRedirect(0x2dc8f80, 0x2838978, 0x1, 0x2, 0x2c436e0, 0x27)
	/home/pi/go/src/github.com/disclose/diosts/pkg/securitytxt/domain.go:190 +0x12c
net/http.(*Client).checkRedirect(0x207ab40, 0x2dc8f80, 0x2838978, 0x1, 0x2, 0x0, 0x600001)
	/usr/lib/go-1.11/src/net/http/client.go:416 +0x4c
net/http.(*Client).do(0x207ab40, 0x2dc9380, 0x0, 0x0, 0x0)
	/usr/lib/go-1.11/src/net/http/client.go:607 +0x738
net/http.(*Client).Do(0x207ab40, 0x2dc9380, 0x2126f30, 0x27, 0x0)
	/usr/lib/go-1.11/src/net/http/client.go:509 +0x24
net/http.(*Client).Get(0x207ab40, 0x2126f30, 0x27, 0x2126f30, 0x27, 0x0)
	/usr/lib/go-1.11/src/net/http/client.go:398 +0x7c
github.com/disclose/diosts/pkg/securitytxt.(*DomainClient).GetBody(0x20787d0, 0x2126f30, 0x27, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/pi/go/src/github.com/disclose/diosts/pkg/securitytxt/domain.go:143 +0x94
github.com/disclose/diosts/pkg/securitytxt.(*DomainClient).GetDomainBody(0x20787d0, 0x2304f50, 0x7, 0x1)
	/home/pi/go/src/github.com/disclose/diosts/pkg/securitytxt/domain.go:114 +0x15c
github.com/disclose/diosts/pkg/securitytxt.(*DomainClient).GetSecurityTxt(0x20787d0, 0x2304f50, 0x7, 0x0, 0x0, 0x0)
	/home/pi/go/src/github.com/disclose/diosts/pkg/securitytxt/domain.go:88 +0xcc
github.com/disclose/diosts/internal/app/run.(*WorkerPool).work(0x207ab60, 0x208e200)
	/home/pi/go/src/github.com/disclose/diosts/internal/app/run/workerpool.go:52 +0xac
created by github.com/disclose/diosts/internal/app/run.(*WorkerPool).Run
	/home/pi/go/src/github.com/disclose/diosts/internal/app/run/workerpool.go:34 +0x54
```

It's a redirect to a URL with the hostname ending with ".". Technically valid, but not expected. Will fix.

```
{"level":"debug","url":"http://cort.as/.well-known/security.txt","time":"2020-08-27T21:39:51-07:00","message":"retrieving"}
{"level":"debug","from":"http://cort.as/.well-known/security.txt","to":"http://cortas.elpais.com./.well-known/security.txt","time":"2020-08-27T21:39:52-07:00","message":"redirecting"}
panic: runtime error: index out of range [0] with length 0
```

--- a/cmd/diosts/commands/root.go
+++ b/cmd/diosts/commands/root.go
@@ -30,6 +30,12 @@ func init() {
 		run.DefaultConfig.NumThreads,
 		"Number of concurrent scraping threads",
 	)
+
+	rootCmd.Flags().BoolVar(&runConfig.SecurityTxt.StrictRedirect,
+		"strict-redirect",
+		false,
+		"Only allow redirects to same base domain",
+	)
 }
 
 func initApp() {

--- a/pkg/securitytxt/config.go
+++ b/pkg/securitytxt/config.go
@@ -9,6 +9,7 @@ type Config struct {
 	RequestTimeout time.Duration
 	TLSHandshakeTimeout time.Duration
 	Insecure bool
+	StrictRedirect bool
 }
 
 var DefaultConfig = Config{
@@ -16,6 +17,7 @@ var DefaultConfig = Config{
 	RequestTimeout: 10 * time.Second,
 	TLSHandshakeTimeout: 5 * time.Second,
 	Insecure: false,
+	StrictRedirect: false,
 }
 
 func NewConfig() *Config {

--- a/pkg/securitytxt/domain.go
+++ b/pkg/securitytxt/domain.go
@@ -197,6 +197,11 @@ func checkRedirect(req *http.Request, via []*http.Request) error {
 
 // Get the base domain name
 func baseDomain(domain string) string {
+	// Remove trailing "." - technically valid, but not needed in this case
+	if domain[len(domain) - 1] == byte('.') {
+		domain = domain[:len(domain) - 2]
+	}
+
 	splits := strings.Split(domain, ".")
 
 	// This is just weird, but ok - probably ipv6 address

--- a/pkg/securitytxt/domain.go
+++ b/pkg/securitytxt/domain.go
@@ -171,7 +171,7 @@ func (c *DomainClient) GetBody(url string) ([]byte, error) {
 // Make sure we don't leave this domain - always log something
 func (c *DomainClient) checkRedirect(req *http.Request, via []*http.Request) error {
 	from := via[len(via) - 1]
-	log.Debug().Str("from", from.URL.String()).Str("to", req.URL.String()).Msg("redirecting")
+	log.Info().Str("from", from.URL.String()).Str("to", req.URL.String()).Msg("redirecting")
 
 	if c.Config.StrictRedirect {
 		fromHost := baseDomain(from.URL.Hostname())

--- a/pkg/securitytxt/domain.go
+++ b/pkg/securitytxt/domain.go
@@ -57,8 +57,12 @@ func NewDomainClient(config *Config) (*DomainClient, error) {
 		KeepAlive: -1,
 	}
 
+	c := &DomainClient{
+		Config: config,
+	}
+
 	client := &http.Client{
-		CheckRedirect: checkRedirect,
+		CheckRedirect: c.checkRedirect,
 		Transport: &http.Transport{
 			TLSClientConfig: &tls.Config{
 				InsecureSkipVerify: config.Insecure,
@@ -71,10 +75,7 @@ func NewDomainClient(config *Config) (*DomainClient, error) {
 		Timeout: config.RequestTimeout,
 	}
 
-	c := &DomainClient{
-		Config: config,
-		client: client,
-	}
+	c.client = client
 
 	return c, nil
 }
@@ -167,6 +168,22 @@ func (c *DomainClient) GetBody(url string) ([]byte, error) {
 	return body, err
 }
 
+// Make sure we don't leave this domain - always log something
+func (c *DomainClient) checkRedirect(req *http.Request, via []*http.Request) error {
+	from := via[len(via) - 1]
+	log.Debug().Str("from", from.URL.String()).Str("to", req.URL.String()).Msg("redirecting")
+
+	if c.Config.StrictRedirect {
+		fromHost := baseDomain(from.URL.Hostname())
+		toHost := baseDomain(req.URL.Hostname())
+		if fromHost != toHost {
+			return NewRedirectError(fromHost, toHost)
+		}
+	}
+
+	return nil
+}
+
 // Get bare domain from input
 func stripDomain(domain string) (string) {
 	// Check for schema, strip if present
@@ -179,20 +196,6 @@ func stripDomain(domain string) (string) {
 	domain = splits[0]
 
 	return domain
-}
-
-// Make sure we don't leave this domain - always log something
-func checkRedirect(req *http.Request, via []*http.Request) error {
-	from := via[len(via) - 1]
-	log.Debug().Str("from", from.URL.String()).Str("to", req.URL.String()).Msg("redirecting")
-
-	fromHost := baseDomain(from.URL.Hostname())
-	toHost := baseDomain(req.URL.Hostname())
-	if fromHost != toHost {
-		return NewRedirectError(fromHost, toHost)
-	}
-
-	return nil
 }
 
 // Get the base domain name

